sklearn/cluster/_optics.py

```diff
@@ -285,7 +288,10 @@ def fit(self, X, y=None):
             )
             warnings.warn(msg, DataConversionWarning)
 
-        X = self._validate_data(X, dtype=dtype)
+        X = self._validate_data(X, dtype=dtype, accept_sparse="csr")
+        if self.metric == "precomputed" and issparse(X):
+            # Set each diagonal to an explicit value so each point is its own neighbor
+            X.setdiag(X.diagonal())
```

Comment:
Scipy complains that this is inefficient for CSR format. Should we validate with ``accept_sparse="lil" if self.metric == "precomputed" else "csr"`` and then convert to csr ?

Reply:
Interestingly enough, it seems as though `lil` (as well as `dok`) don't seem to be able to accept explicit zero-values which is what needs to happen to have the graph built correctly. We could manually do it, but that'll probably end up even worse than the current implementation for any decently sized matrix. 

I may just be missing an obvious way to do this since I'm unfamiliar with manipulating scipy sparse matrices, but afaik this is a necessary evil right now.

Reply:
I do not think any estimators with `metric="precomputed"` support precomputed sparse data. In practice, I do not think precomputed distances are usually sparse.

With that in mind, I would be happy with raising an error for `metric == "precomputed"` and sparse data.

Reply:
maybe we can filter this warning and add a small comment.

Reply:
I am okay with filtering, since we do the same thing in DBSCAN:

https://github.com/scikit-learn/scikit-learn/blob/f9a82d28e7f0f0b1f274acad9607eaa38fb12bc8/sklearn/cluster/_dbscan.py#L355-L360